File: problems/kotlin/3.longest-substring-without-repeating-characters.kt
/*
 * @lc app=leetcode id=3 lang=kotlin
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var tmp = HashSet<Char>()
        var max = 0
        var i = 0
        var j = 0

        while (j < s.length) {
            if (tmp.contains(s[j])) {
                tmp.remove(s[i])
                i++
            } else {
                tmp.add(s[j])
                j++
                max = Math.max(max, j - i)
            }
        }
        return max
    }
}
// @lc code=end

File: problems/python/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
from typing import Set

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen: Set[str] = set()
        max_count, i, j = 0, 0, 0

        while j < len(s):
            if s[j] in seen:
                seen.remove(s[i])
                i+=1
            else:
                seen.add(s[j])
                j+=1
                max_count = max(max_count, j - i)

        return max_count
# @lc code=end

